### ISSUE TITLE
Let `await` hang the isolate + observatory support.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#23797]
Currently the VM stops an isolate when there are no queued events, scheduled timers or open receive ports. That means that code like:
```dart
main() async {
   await foo();
   print("DONE");
}
```
may terminate at the await if the `foo` code returns a future with nobody scheduled to complete it. That can happen through any number of bugs, but it is hard to debug when the VM just terminates. The code looks like direct-style code, and the user expects it to continue after the await.
Similarly it may fail to run finally blocks:
```dart
main() async {
  try {
     await foo();
  } finally {
     print("DONE");
  }
}
```
This can also terminate at the await, which is even less expected. It's not absolutely *wrong* because the code does't actually skip the finally, it just kills the isolate. 

In any case, it's unexpected behavior and hard to debug.

How about letting all pending await calls be registered by the isolate, and visible using the observatory, and not terminate the isolate while an await is waiting. That would change the behavior from early termination to look like a deadlock (which it most likely is). The user can then use the observatory to see in which places in async functions the isolate is currently waiting which will aid debugging - it's easy to see why "DONE" is not being printed if the await on the previous line is blocked.

**Comment by sethladd** [comment]
cc @johnmccutchan

**Comment by johnmccutchan** [comment]
@rmacnak-google

**Comment by turnidge** [comment]
This bug is a two part bug.  One part is an Observatory feature request, the other is a kind of await feature request asking for different vm termination conditions.  I'm adding @mhausner to the bug to comment on the latter.

**Comment by mhausner** [comment]
I think Lasse has a good point here. I'll look into this.

It wasn't clear to me immediately how you create a Future that isn't scheduled for completion, but for the record:

main() async {
   await new Completer().future;
   print("DONE");
}

**Comment by mhausner** [comment]
On second thought, is "hanging" the isolate the right thing to do? And why only for futures that are being awaited? Why not all uncompleted futures, or even uncompleted futures that may have been garbage collected because nobody has a reference to them anymore?

Ignoring the Observatory feature wish, would it be a good enough approximation to make sure the future returned by main() completes?

**Comment by lrhn** [comment]
The reason for making this only apply to "await" expressions is that they look like straight-line code, and users will expect straight-line code behavior from it. It doesn't look like code that is split into separate chunks that will each eventually (maybe) be called from the event loop, so actually behaving like that is surprising - function terminates in the middle without any errors being printed - nothing suggests that `foo(f) async {print("before");await f;print("after");}` should be able to print "before", but neither print "after" nor throw - there is no control flow point in the middle that allows it to suddently exit (and omit finally blocks on top of that).

This is proposing that async/await code should act like it is a normal function call being blocked at the await. Until that function call terminates, the program isn't "done", it's just stuck. That's what the code is was made to look like, and I think its behavior should follow that.

For an end user, the difference will be between hanging and terminating silently. The former is easier to both explain and debug.

Futures that are just used "normally" (without await) are much more obviously asynchronous and using the event loop, and it's less surprising if they never get completed. Also, there are probably quite a lot of futures that are never completed. If nobody waits on them, that's not a problem. It only becomes a problem with await because someone, inside (what looks like) a running function call, is stuck waiting for that future.

It's not necessarily enough to check that a future returned by main completes - for one thing, there might not be one, you could call an async function from a sync main and ignore the result.

**Comment by lrhn** [comment]
Changing this to Area-VM since the main feature being requested is not exiting on a blocked `await` expression. That's worth doing in any case. 
Having the observatory help debugging a stuck program would be a nice-to-have feature, but it depends on the former change.

**Comment by yjbanov** [comment]
This seems like a bug to me. I just hit it when I tried to complete a `Completer` from a VM service extension. Here's the smallest repro case I could think of:

```dart
import 'dart:async';
import 'dart:developer';

Future<Null> main() async {
  try {
    print('before');
    var c = new Completer();
    registerExtension('ext.complete', (_, __) {
      c.complete();
    });
    await c.future;
  } finally {
    print('after');
  }
}
```

You never see the "after" message despite the fact that the completer may be completed from the extension method. The process just quits.

Is there a workaround for this?

**Comment by rmacnak-google** [comment]
You could keep the VM alive by opening a receive port when you create the completer and closing the port when you complete it.

**Comment by yjbanov** [comment]
@rmacnak-google's workaround works. Thanks!

Also, I no longer think that it's a bug. VM service extensions by themselves should not alter the behavior of the program. However, the `Completer` API could have a formal way letting the developer indicate that the `Completer` is expected to complete before the program ends. Strawman:

```dart
new Completer(waitUntilComplete: true); // false by default
```

**Comment by tvolkert** [comment]
This happens in any main() program, not just VM extensions, so I think it is a bug.  See reduced case in  #28769

**Comment by kevmoo** [comment]
This has been brought up a few times by customers, without a clear work-around.

**Comment by kevmoo** [comment]
I think the suggestion from @yjbanov is a pretty decent idea.

**Comment by filiph** [comment]
FYI: There's a recent (and unresolved) question about this on StackOverflow:
http://stackoverflow.com/questions/42261143/is-it-possible-to-detect-dead-completers-in-dart

**Comment by tvolkert** [comment]
This just came up again internally in google as a source of confusion.

It sounds like there's agreement that we should not exit on a blocked `await` expression.  Two associated questions:

1. Given that this would be a functional change, what process do we need to go through to socialize and make the change?

2. How difficult is the change?

**Comment by MichaelRFairhurst** [comment]
Also worth noting I think, that our goal was to have the program run forever.

Was going to suggest in code review to do `await new Completer().future`, at the end of `main` but instead we can do

```
while (true) {
  await new Future.delayed(const Duration(seconds: 5));
}
```

or, to avoid the loop:

```
Zone.current.createPeriodicTimer(const Duration(seconds: 5), (){});
```

So I guess we have a workaround, but this definitely threw me for a loop. (In fact, I still don't really understand why this doesn't work)

**Comment by lrhn** [comment]
The typical work-around to keep an isolate alive is `var keepAlive = new ReceivePort();`. 
Receive-ports keep the isolate alive because they can potentially receive events, and they are not garbage collected because we don't have cross-isolate GC that can see that no corresponding send-port is alive.

**Comment by shyndman** [comment]
Hello,

Is P2 appropriate here? If I were a newer user to Dart and encountered this, it wouldn't leave a good impression. The current behaviour can be quite difficult to debug.

**Comment by anidotnet** [comment]
Is there any solution or direction yet? It's been long since this issue was opened. I faced similar problem with await for loop with streams as below.

```dart
import 'package:rxdart/rxdart.dart';

void main() async {
  ReplaySubject<int> subject = ReplaySubject<int>();
  subject.add(1);
  Future.delayed(Duration(seconds: 1), () {
    subject.add(2);
  });
  subject.add(3);

  await for (var value in subject) {
    print(value);
  }

  print("Completed");
}

```

The above code never prints "Completed", which is very confusing to me, who is coming from Java/C# world. Generally **await** means, wait for the async operation to complete and then move forward, which is not the case here it seems.

On the contrary when I use `forEach()` method on stream it behaves as expected.

```dart
import 'package:rxdart/rxdart.dart';

void main() async {
  ReplaySubject<int> subject = ReplaySubject<int>();
  subject.add(1);
  Future.delayed(Duration(seconds: 1), () {
    subject.add(2);
  });
  subject.add(3);

  subject.forEach(print);

  print("Completed");
}
```

So my question is, do we have a dart idiomatic way of doing the following things?

- invoke an async function
- wait for the above async function to complete and return result or throw error
- move forward with further statements.

Any pointer would be helpful. Currently I am a bit hesitant to use dart/flutter for big project as a issue like this is very hard to debug in large program.

**Comment by lrhn** [comment]
@anidotnet The reason `forEach` "works" is that you forgot to await the future it returned. If you had done that, as is usually recommended, it would work just like `await for`.

The idiomatic way to wait for an async operation to complete is to `await` the future it returns. 
You call it, it returns a future (that's what makes it an async function). You await the future. When that completes, you move forward with the code after the `await`.
The code won't continue past that `await` until the operation is done (meaning it completes the future it returned). The problem addressed here is that it's possible for a future to never complete, or similarly, a stream to never end.

The reason your code doesn't end is that you make a loop to wait for *all* events of a stream, and then you never close the stream, so the loop keeps waiting. That's reasonable behavior. Not useful, but it's all it can do.

What this issue is suggesting is that the program won't *terminate* while you are waiting. Your code wont continue past the `await`/`await for` until the operation is done/stream is closed, but if the operation hangs, which an asynchronous function can do by simply doing nothing and never completing its future, or a stream never being closed, the `await`/`await for` never completes. The code after the `await` never runs, ever. 
The program may still terminate, because being stuck at an await is not special, it's just a callback stored somewhere, which is not being called. That alone does not keep the program alive when there is no other scheduled work to do.

**Comment by anidotnet** [comment]
Thanks for the detailed explanations @lrhn. Based on your comment, I have made following tweak in the code and it started working as expected.

```dart
import 'package:rxdart/rxdart.dart';

void main() async {
  ReplaySubject<int> subject = ReplaySubject<int>();
  subject.add(1);
  Future.delayed(Duration(seconds: 1), () {
    subject.add(2);
    if (!subject.isClosed) {
      subject.close();
    }
  });
  subject.add(3);

  await for (var value in subject) {
    print(value);
  }

  print("Completed");
}
```
Thanks for the help.
